File: SiriusQuality/SiriusQuality-ET/src/apsim/SiriusQuality_ET/Conductance.cs
using System;
using System.Collections.Generic;
using System.Linq;    
using Models.Core;   
namespace Models.Crop2ML;

/// <summary>
///- Name: Conductance -Version: 1.0, -Time step: 1
///- Description:
///            * Title: Conductance Model
///            * Authors: Peter D. Jamieson, Glen S. Francis, Derick R. Wilson, Robert J. Martin
///            * Reference: https://doi.org/10.1016/0168-1923(94)02214-5
///
///            * Institution: New Zealand Institute for Crop and Food Research Ltd.,
///New Zealand Institute for Crop and Food Research Ltd.,
///New Zealand Institute for Crop and Food Research Ltd.,
///New Zealand Institute for Crop and Food Research Ltd.
///
///            * ExtendedDescription: The boundary layer conductance is expressed as the wind speed profile above the
///canopy and the canopy structure. The approach does not take into account buoyancy
///effects.
///
///            * ShortDescription: The boundary layer conductance is expressed as the wind speed profile above the
///canopy and the canopy structure. The approach does not take into account buoyancy
///effects.
///
///- inputs:
///            * name: d
///                          ** description : corresponding to 2/3. This is multiplied to the crop heigth for calculating the zero plane displacement height, FAO
///                          ** inputtype : parameter
///                          ** parametercategory : constant
///                          ** datatype : DOUBLE
///                          ** max : 1
///                          ** min : 0
///                          ** default : 0.67
///                          ** unit : dimensionless
///            * name: heightWeatherMeasurements
///                          ** description : reference height of wind and humidity measurements
///                          ** inputtype : parameter
///                          ** parametercategory : soil
///                          ** datatype : DOUBLE
///                          ** max : 10
///                          ** min : 0
///                          ** default : 2
///                          ** unit : m
///            * name: plantHeight
///                          ** description : plant Height
///                          ** inputtype : variable
///                          ** variablecategory : auxiliary
///                          ** datatype : DOUBLE
///                          ** max : 1000
///                          ** min : 0
///                          ** default : 0
///                          ** unit : mm
///            * name: zh
///                          ** description : roughness length governing transfer of heat and vapour, FAO
///                          ** inputtype : parameter
///                          ** parametercategory : constant
///                          ** datatype : DOUBLE
///                          ** max : 1
///                          ** min : 0
///                          ** default : 0.013
///                          ** unit : m
///            * name: zm
///                          ** description : roughness length governing momentum transfer, FAO
///                          ** inputtype : parameter
///                          ** parametercategory : constant
///                          ** datatype : DOUBLE
///                          ** max : 1
///                          ** min : 0
///                          ** default : 0.13
///                          ** unit : m
///            * name: vonKarman
///                          ** description : von Karman constant
///                          ** inputtype : parameter
///                          ** parametercategory : constant
///                          ** datatype : DOUBLE
///                          ** max : 1
///                          ** min : 0
///                          ** default : 0.42
///                          ** unit : dimensionless
///            * name: ih
///                          ** description : hour of the day if the component is hourly, -999 if the component is daily
///                          ** inputtype : variable
///                          ** parametercategory : state
///                          ** datatype : INT
///                          ** max : 24
///                          ** min : 999
///                          ** default : 999
///                          ** unit : 
///            * name: wind
///                          ** description : wind
///                          ** inputtype : variable
///                          ** variablecategory : auxiliary
///                          ** datatype : DOUBLE
///                          ** max : 1000000
///                          ** min : 0
///                          ** default : 124000
///                          ** unit : m/d
///- outputs:
///            * name: conductance
///                          ** description : the boundary layer conductance
///                          ** datatype : DOUBLE
///                          ** variablecategory : state
///                          ** max : 10000
///                          ** min : 0
///                          ** unit : m/d
/// </summary>
public class Conductance
{

    private double _d;
    /// <summary>
    /// Gets and sets the corresponding to 2/3. This is multiplied to the crop heigth for calculating the zero plane displacement height, FAO
    /// </summary>
    [Description("corresponding to 2/3. This is multiplied to the crop heigth for calculating the zero plane displacement height, FAO")] 
    [Units("dimensionless")] 
    //[Crop2ML(datatype="DOUBLE", min=0, max=1, default=0.67, parametercategory=constant, inputtype="parameter")] 
    public double d
    {
        get { return this._d; }
        set { this._d= value; } 
    }

    private double _heightWeatherMeasurements;
    /// <summary>
    /// Gets and sets the reference height of wind and humidity measurements
    /// </summary>
    [Description("reference height of wind and humidity measurements")] 
    [Units("m")] 
    //[Crop2ML(datatype="DOUBLE", min=0, max=10, default=2, parametercategory=soil, inputtype="parameter")] 
    public double heightWeatherMeasurements
    {
        get { return this._heightWeatherMeasurements; }
        set { this._heightWeatherMeasurements= value; } 
    }

    private double _zh;
    /// <summary>
    /// Gets and sets the roughness length governing transfer of heat and vapour, FAO
    /// </summary>
    [Description("roughness length governing transfer of heat and vapour, FAO")] 
    [Units("m")] 
    //[Crop2ML(datatype="DOUBLE", min=0, max=1, default=0.013, parametercategory=constant, inputtype="parameter")] 
    public double zh
    {
        get { return this._zh; }
        set { this._zh= value; } 
    }

    private double _zm;
    /// <summary>
    /// Gets and sets the roughness length governing momentum transfer, FAO
    /// </summary>
    [Description("roughness length governing momentum transfer, FAO")] 
    [Units("m")] 
    //[Crop2ML(datatype="DOUBLE", min=0, max=1, default=0.13, parametercategory=constant, inputtype="parameter")] 
    public double zm
    {
        get { return this._zm; }
        set { this._zm= value; } 
    }

    private double _vonKarman;
    /// <summary>
    /// Gets and sets the von Karman constant
    /// </summary>
    [Description("von Karman constant")] 
    [Units("dimensionless")] 
    //[Crop2ML(datatype="DOUBLE", min=0, max=1, default=0.42, parametercategory=constant, inputtype="parameter")] 
    public double vonKarman
    {
        get { return this._vonKarman; }
        set { this._vonKarman= value; } 
    }

    private int _ih;
    /// <summary>
    /// Gets and sets the hour of the day if the component is hourly, -999 if the component is daily
    /// </summary>
    [Description("hour of the day if the component is hourly, -999 if the component is daily")] 
    [Units("")] 
    //[Crop2ML(datatype="INT", min=999, max=24, default=999, parametercategory=state, inputtype="parameter")] 
    public int ih
    {
        get { return this._ih; }
        set { this._ih= value; } 
    }

    
    /// <summary>
    /// Constructor of the Conductance component")
    /// </summary>  
    public Conductance() { }
    
    /// <summary>
    /// Algorithm of the Conductance component
    /// </summary>
    public void  CalculateModel(EnergyBalanceCompositeState s, EnergyBalanceCompositeState s1, EnergyBalanceCompositeRate r, EnergyBalanceCompositeAuxiliary a, EnergyBalanceCompositeExogenous ex)
    {
        double plantHeight = a.plantHeight;
        double wind = a.wind;
        double conductance;
        double h;
        double clim;
        clim = 0.10;
        if (ih != -999)
        {
            clim = 36.00;
        }
        h = Math.Max(10.00, plantHeight) / 100.00;
        conductance = wind * Math.Pow(vonKarman, 2) / (Math.Log((heightWeatherMeasurements - (d * h)) / (zm * h)) * Math.Log((heightWeatherMeasurements - (d * h)) / (zh * h)));
        conductance = Math.Max(clim, conductance);
        s.conductance= conductance;
    }
}

File: SiriusQuality/SiriusQuality-EnergyBalance_Original/src/simplace/SiriusQuality_EnergyBalance/Conductance.java
package net.simplace.sim.components.SiriusQuality-EnergyBalance;
import  java.io.*;
import  java.util.*;
import java.text.ParseException;
import java.text.SimpleDateFormat;
import java.time.LocalDateTime;
import net.simplace.sim.model.FWSimComponent;
import net.simplace.sim.util.FWSimVarMap;
import net.simplace.sim.util.FWSimVariable;
import net.simplace.sim.util.FWSimVariable.CONTENT_TYPE;
import net.simplace.sim.util.FWSimVariable.DATA_TYPE;
import org.jdom2.Element;


public class Conductance extends FWSimComponent
{
    private FWSimVariable<Integer> ih;
    private FWSimVariable<Double> wind;
    private FWSimVariable<Double> heightWeatherMeasurements;
    private FWSimVariable<Double> vonKarman;
    private FWSimVariable<Double> plantHeight;
    private FWSimVariable<Double> zh;
    private FWSimVariable<Double> zm;
    private FWSimVariable<Double> d;
    private FWSimVariable<Double> conductance;

    public Conductance(String aName, HashMap<String, FWSimVariable<?>> aFieldMap, HashMap<String, String> aInputMap, Element aSimComponentElement, FWSimVarMap aVarMap, int aOrderNumber)
    {
        super(aName, aFieldMap, aInputMap, aSimComponentElement, aVarMap, aOrderNumber);
    }

    public Conductance(){
        super();
    }

    @Override
    public HashMap<String, FWSimVariable<?>> createVariables()
    {
        addVariable(FWSimVariable.createSimVariable("ih", "hour of the day if the component is hourly, -999 if the component is daily", DATA_TYPE.INT, CONTENT_TYPE.constant,"", 999, 24, 999, this));
        addVariable(FWSimVariable.createSimVariable("wind", "wind", DATA_TYPE.DOUBLE, CONTENT_TYPE.input,"m/d", 0, 1000000, 124000, this));
        addVariable(FWSimVariable.createSimVariable("heightWeatherMeasurements", "reference height of wind and humidity measurements", DATA_TYPE.DOUBLE, CONTENT_TYPE.constant,"m", 0, 10, 2, this));
        addVariable(FWSimVariable.createSimVariable("vonKarman", "von Karman constant", DATA_TYPE.DOUBLE, CONTENT_TYPE.constant,"dimensionless", 0, 1, 0.42, this));
        addVariable(FWSimVariable.createSimVariable("plantHeight", "plant Height", DATA_TYPE.DOUBLE, CONTENT_TYPE.input,"mm", 0, 1000, 0, this));
        addVariable(FWSimVariable.createSimVariable("zh", "roughness length governing transfer of heat and vapour, FAO", DATA_TYPE.DOUBLE, CONTENT_TYPE.constant,"m", 0, 1, 0.013, this));
        addVariable(FWSimVariable.createSimVariable("zm", "roughness length governing momentum transfer, FAO", DATA_TYPE.DOUBLE, CONTENT_TYPE.constant,"m", 0, 1, 0.13, this));
        addVariable(FWSimVariable.createSimVariable("d", "corresponding to 2/3. This is multiplied to the crop heigth for calculating the zero plane displacement height, FAO", DATA_TYPE.DOUBLE, CONTENT_TYPE.constant,"dimensionless", 0, 1, 0.67, this));
        addVariable(FWSimVariable.createSimVariable("conductance", "the boundary layer conductance", DATA_TYPE.DOUBLE, CONTENT_TYPE.state,"m/d", 0, 10000, null, this));

        return iFieldMap;
    }
    @Override
    protected void process()
    {
        Integer t_ih = ih.getValue();
        double t_wind = wind.getValue();
        double t_heightWeatherMeasurements = heightWeatherMeasurements.getValue();
        double t_vonKarman = vonKarman.getValue();
        double t_plantHeight = plantHeight.getValue();
        double t_zh = zh.getValue();
        double t_zm = zm.getValue();
        double t_d = d.getValue();
        double t_conductance = conductance.getDefault();
        double h;
        double clim;
        clim = 0.10d;
        if (t_ih != -999)
        {
            clim = 36.00d;
        }
        h = Math.max(10.00d, t_plantHeight) / 100.00d;
        t_conductance = t_wind * Math.pow(t_vonKarman, 2) / (Math.log((t_heightWeatherMeasurements - (t_d * h)) / (t_zm * h)) * Math.log((t_heightWeatherMeasurements - (t_d * h)) / (t_zh * h)));
        t_conductance = Math.max(clim, t_conductance);
        conductance.setValue(t_conductance, this);
    }

    @Override
    protected void init()
    {
    }
    public HashMap<String, FWSimVariable<?>> fillTestVariables(int aParamIndex, TEST_STATE aDefineOrCheck)
    {
        return iFieldMap;
    }

    @Override
    protected FWSimComponent clone(FWSimVarMap aVarMap)
    {
        return new Conductance(iName, iFieldMap, iInputMap, iSimComponentElement, aVarMap, iOrderNumber);
    }
}